File: harness/lib/observability/python/metrics.py
"""
Métricas HTTP com prometheus_client.
"""
from __future__ import annotations

import time
from typing import Any, Callable

from prometheus_client import REGISTRY, Counter, Histogram, generate_latest, CONTENT_TYPE_LATEST

http_requests_total = Counter(
    "http_requests_total",
    "Total number of HTTP requests",
    ["method", "route", "status_code"],
)

http_request_duration_seconds = Histogram(
    "http_request_duration_seconds",
    "HTTP request duration in seconds",
    ["method", "route", "status_code"],
    buckets=[0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0],
)


def metrics_content() -> tuple[bytes, str]:
    """Retorna o conteúdo das métricas no formato Prometheus."""
    return generate_latest(REGISTRY), CONTENT_TYPE_LATEST


# ── FastAPI route ──────────────────────────────────────────────────────────────
try:
    from fastapi import APIRouter, Response

    metrics_router = APIRouter()

    @metrics_router.get("/metrics")
    async def prometheus_metrics() -> Response:
        content, content_type = metrics_content()
        return Response(content=content, media_type=content_type)

except ImportError:
    pass

File: harness/lib/observability/node/metrics.ts
import { collectDefaultMetrics as promCollectDefaultMetrics, Counter, Histogram, Registry } from 'prom-client';

export const registry = new Registry();

export const httpRequestsTotal = new Counter({
  name: 'http_requests_total',
  help: 'Total number of HTTP requests',
  labelNames: ['method', 'route', 'status_code'],
  registers: [registry],
});

export const httpRequestDurationMs = new Histogram({
  name: 'http_request_duration_ms',
  help: 'HTTP request duration in milliseconds',
  labelNames: ['method', 'route', 'status_code'],
  buckets: [5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000],
  registers: [registry],
});

export const activeRequests = new Counter({
  name: 'http_active_requests',
  help: 'Number of currently active HTTP requests',
  registers: [registry],
});

export function collectDefaultMetrics(): void {
  promCollectDefaultMetrics({ register: registry });
}

/**
 * Express middleware que registra métricas HTTP automaticamente.
 * Uso: app.use(metricsMiddleware)
 */
export function metricsMiddleware(
  req: { method: string; path?: string; route?: { path: string } },
  res: { statusCode: number; on(event: string, fn: () => void): void },
  next: () => void,
): void {
  const start = Date.now();
  activeRequests.inc();

  res.on('finish', () => {
    const route = req.route?.path ?? req.path ?? 'unknown';
    const labels = {
      method: req.method,
      route,
      status_code: String(res.statusCode),
    };
    httpRequestsTotal.inc(labels);
    httpRequestDurationMs.observe(labels, Date.now() - start);
    activeRequests.inc(-1);
  });

  next();
}
